featurebyte/query_graph/node/function.py

```diff
@@ -0,0 +1,228 @@
+    CodeGenerationContext,
+    ExpressionStr,
+    InfoDict,
+    RightHandSide,
+    StatementT,
+    ValueStr,
+    VariableNameGenerator,
+    VarNameExpressionInfo,
+)
+from featurebyte.query_graph.node.scalar import ValueParameterType
+
+
+class BaseFunctionParameterInput(BaseModel):
+    """BaseFunctionParameterInput class"""
+
+    dtype: DBVarType
+
+    @abstractmethod
+    def get_column_args(self) -> List[Optional[str]]:
+        """
+        Get column arguments (column name or None) for the function parameter input.
+        If the input is a value, return empty list.
+
+        Returns
+        _______
+        List[Optional[str]]
+        """
+
+    @abstractmethod
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        """
+        Get SDK function argument for the function parameter input.
+
+        Parameters
+        ----------
+        node_inputs : List[VarNameExpressionInfo]
+            List of node inputs.
+        node_input_index : int
+            The index of the node input that the function parameter input is associated with.
+
+        Returns
+        -------
+        Tuple[RightHandSide, int]
+            Tuple of SDK function argument and the updated node input index.
+        """
+
+
+class ValueFunctionParameterInput(BaseFunctionParameterInput):
+    """ValueFunctionParameterInput class"""
+
+    value: Optional[ValueParameterType]
+    input_form: Literal[FunctionParameterInputForm.VALUE] = Field(
+        FunctionParameterInputForm.VALUE, const=True
+    )
+
+    def get_column_args(self) -> List[Optional[str]]:
+        return []
+
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        # do not increment node_input_index as this is a value input, it should not increment the node_input_index
+        return ValueStr.create(self.value), node_input_index
+
+
+class ColumnFunctionParameterInput(BaseFunctionParameterInput):
+    """ColumnFunctionParameterInput class"""
+
+    column_name: Optional[str]
+    input_form: Literal[FunctionParameterInputForm.COLUMN] = Field(
+        FunctionParameterInputForm.COLUMN, const=True
+    )
+
+    def get_column_args(self) -> List[Optional[str]]:
+        return [self.column_name]
+
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        # InfoDict is not expected here as it should be used only in (ConditionNode - AssignNode) structure
+        value = node_inputs[node_input_index]
+        assert not isinstance(value, InfoDict), "Unexpected InfoDict type"
+        # increment node_input_index as this is a column input, it should increment the node_input_index
+        return value, node_input_index + 1
+
+
+FunctionParameterInput = Annotated[
```

Comment:
curious - is this type needed? could `function_parameters` be of the type `List[BaseFunctionParameterInput]` instead?

Reply:
This is needed. Without this, the input parameter will be dropped after deserialization.

Reply:
got it, thanks!

Reply:
For example, `BaseFunctionParameterInput` only contains `dtype`. Those additional attributes that are specific to value or column will be dropped after deserialization (since the expected type is `BaseFunctionParameterInput` and it does not contain these parameters).